crates/sui-json-rpc/src/indexer_api.rs

```diff
@@ -371,26 +370,103 @@ impl<R: ReadApiServer> IndexerApiServer for IndexerApi<R> {
     #[instrument(skip(self))]
     async fn resolve_name_service_address(&self, name: String) -> RpcResult<Option<SuiAddress>> {
         with_tracing!(async move {
-            let domain = Domain::from_str(&name).map_err(|e| {
-                Error::UnexpectedError(format!(
-                    "Failed to parse NameService Domain with error: {:?}",
-                    e
-                ))
-            })?;
+            let requested_domain = name.parse::<Domain>().map_err(Error::from)?;
```

Comment:
This looks OK to me from a correctness perspective, but ...is this not a lot of work to be doing every time we need to resolve a name?  If we want to live in a world where most of the objects and addresses people regularly interact with come with a domain name, this is not likely to scale -- there are multiple sequential accesses to the DB on this path.  Do we have a plan for that eventuality?

Reply:
So, here, the worst case scenario is 2 sequential reads to the database (for a subdomain resolution):
1. Get the SLD name
2. Get subdomain (parent / child), which relies on the 1st result (that holds the namespace ID).

I am not sure if there's a way to eventually get down to a single db read, maybe through a special indexing table?

Reply:
Do our domain names have a TTL, so that RPCs can cache them in memory?

Reply:
They do not.
What we can cache though, which can bring down subdomain lookups (after a single one has been made under a namespace) to a single db query, is the `Namespace_id` for subdomain lookups, until the expiration of the parent (which is in the months/years range).

Reply:
It sounds like the scaling topic is not something we've fully explored yet, is that right? If so, we should kick off a discussion about that, because even in the case where we can cache the IDs of tables holding sub-domains, if we're expecting every regularly accessed package and object to have a domain, we're going to see some quite painful read amplification.

Reply:
Going completely away from DB reads, isn't something that has been explored afaik and agreed we should talk about this further! Maybe @bmwill has more context on this (whether this has been explored)